sway-core/src/ir_generation/function.rs

```diff
@@ -869,35 +923,59 @@ impl FnCompiler {
 
         // Jump to the while cond block.
         let cond_block = self.function.create_block(context, Some("while".into()));
-        self.current_block
-            .ins(context)
-            .branch(cond_block, None, None);
+
+        if !self.current_block.is_terminated_by_a_branch(context) {
+            self.current_block
+                .ins(context)
+                .branch(cond_block, None, None);
+        }
 
         // Fill in the body block now, jump unconditionally to the cond block at its end.
         let body_block = self
             .function
             .create_block(context, Some("while_body".into()));
-        self.current_block = body_block;
-        self.compile_code_block(context, ast_while_loop.body)?;
-        self.current_block
-            .ins(context)
-            .branch(cond_block, None, None);
 
         // Create the final block after we're finished with the body.
         let final_block = self
             .function
             .create_block(context, Some("end_while".into()));
 
+        // Keep track of the previous blocks we have to jump to in case of a break or a continue.
+        // This should be `None` if we're not in a loop already or the previous break or continue
+        // destinations for the outer loop that contains the current loop.
+        let prev_block_to_break_to = self.block_to_break_to;
+        let prev_block_to_continue_to = self.block_to_continue_to;
```

Comment:
This feels a bit clumsy (and I balked a bit when I saw `FnCompiler::block_to_break_to` initially) but I can't think of an obviously better technique.  If it wasn't for the `lexical_map` I'd say create a copy of `self` for the recursion, but that doesn't help that much anyway.  Maybe some preprocessing which tags the break or continue with their destinations before compilation, but that's adding unneeded complexity too.

So, nevermind, this is good. :slightly_smiling_face: